qutebrowser/completion/models/listcategory.py

```diff
@@ -0,0 +1,121 @@
+# vim: ft=python fileencoding=utf-8 sts=4 sw=4 et:
+
+# Copyright 2017 Ryan Roden-Corrent (rcorre) <ryan@rcorre.net>
+#
+# This file is part of qutebrowser.
+#
+# qutebrowser is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# qutebrowser is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
+
+"""The base completion model for completion in the command line.
+
+Module attributes:
+    Role: An enum of user defined model roles.
+"""
+
+import re
+
+from PyQt5.QtCore import QSortFilterProxyModel
+from PyQt5.QtGui import QStandardItem, QStandardItemModel
+
+from qutebrowser.utils import qtutils, debug, log
+
+
+class ListCategory(QSortFilterProxyModel):
+
+    """Expose a list of items as a category for the CompletionModel."""
+
+    def __init__(self, name, items, columns_to_filter=None, parent=None):
+        super().__init__(parent)
+        self.name = name
+        self.srcmodel = QStandardItemModel(parent=self)
+        self.pattern = ''
+        self.pattern_re = None
+        self.columns_to_filter = columns_to_filter or [0]
+        for item in items:
+            self.srcmodel.appendRow([QStandardItem(x) for x in item])
+        self.setSourceModel(self.srcmodel)
+
+    def set_pattern(self, val):
+        """Setter for pattern.
+
+        Args:
+            val: The value to set.
+        """
+        self.pattern = val
+        val = re.sub(r' +', r' ', val)  # See #1919
+        val = re.escape(val)
+        val = val.replace(r'\ ', '.*')
+        self.pattern_re = re.compile(val, re.IGNORECASE)
+        self.invalidate()
+        sortcol = 0
+        self.sort(sortcol)
+
+    def filterAcceptsRow(self, row, parent):
```

Comment:
Actually most models only filter the first column, though filtering all non-empty would probably be nice, as would getting rid of `columns_to_filter`